libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs

```diff
@@ -197,5 +208,66 @@ private async Task<Activity> SendToSkillAsync(ITurnContext context, Activity act
 
             return eocActivity;
         }
+
+        private async Task<bool> InterceptOAuthCardsAsync(ITurnContext turnContext, Activity activity, string connectionName, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null && !string.IsNullOrWhiteSpace(connectionName))
+            {
+                foreach (var attachment in activity.Attachments.Where(a => a?.ContentType == OAuthCard.ContentType))
+                {
+                    var oauthCard = ((JObject)attachment.Content).ToObject<OAuthCard>();
+                    if (!string.IsNullOrWhiteSpace(oauthCard.TokenExchangeResource?.Uri))
+                    {
+                        var tokenExchangeProvider = turnContext.Adapter as IExtendedUserTokenProvider;
+                        if (tokenExchangeProvider == null)
+                        {
+                            return false;
```

Comment:
Is this correct? Or should we throw an exception?

Reply:
In other places where adapter is not IExtendedUserTokenProvider, we throw an exception (see OauthPrompt.cs). We may want to be consistent there if it makes sensse in this context.

Reply:
if the developer implemented their own adapter that does not support token exchange, then skill dialog will never work for them if we throw an exception here. The fact that someone wrote an adapter that does not support token exchange is expected and we should follow the behavior of falling back and showing the oauth card

Reply:
If the adapter is the wrong type we throw InvalidOperationExceptions in other dialogs, see OAuthPrompt:

https://github.com/microsoft/botbuilder-dotnet/blob/4a9305e6d7a9238e79140d2e470c040f64ecc5f0/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs#L147-L150

Reply:
we throw in OAuthPrompt dialog, because it's expected that you are using an adapter that supports token exchange and oauth, or else oauthprompt wont work as expected. That is not an expectation here, since you are free to use any adapter that you want and would expect being shown an oauth card even if your adapter does not support token exchange